src/java.base/share/classes/java/util/Locale.java

```diff
@@ -939,29 +939,38 @@ public static Locale getDefault() {
      */
     public static Locale getDefault(Locale.Category category) {
         // do not synchronize this method - see 4071298
-        switch (category) {
-        case DISPLAY:
-            if (defaultDisplayLocale == null) {
-                synchronized(Locale.class) {
-                    if (defaultDisplayLocale == null) {
-                        defaultDisplayLocale = initDefault(category);
-                    }
-                }
+        Objects.requireNonNull(category);
+        if (category == Category.DISPLAY) {
+            Locale loc = defaultDisplayLocale; // volatile read
+            if (loc == null) {
+                loc = getDisplayLocale();
```

Comment:
Just interesting how did you check that the performance difference is because of volatile read, and not because of replacing of the switch by the "if"?

Reply:
I started out with this variant, only removing the double volatile reads:

```
    public static Locale getDefault(Locale.Category category) {
        // do not synchronize this method - see 4071298
        Locale loc;
        switch (category) {
        case DISPLAY:
            loc = defaultDisplayLocale;
            if (loc == null) {
                synchronized(Locale.class) {
                    loc = defaultDisplayLocale;
                    if (loc == null) {
                        loc = defaultDisplayLocale = initDefault(category);
                    }
                }
            }
            return loc;
        case FORMAT:
            loc = defaultFormatLocale;
            if (loc == null) {
                synchronized(Locale.class) {
                    loc = defaultFormatLocale;
                    if (loc == null) {
                        loc = defaultFormatLocale = initDefault(category);
                    }
                }
            }
            return loc;
        default:
            assert false: "Unknown Category";
        }
        return getDefault();
    }
```

Scores were the same:
```
Benchmark                         Mode  Cnt   Score   Error  Units
LocaleDefaults.getDefault         avgt    5  10.045 ± 0.032  ns/op
LocaleDefaults.getDefaultDisplay  avgt    5  11.301 ± 0.053  ns/op
LocaleDefaults.getDefaultFormat   avgt    5  11.303 ± 0.054  ns/op
```

I then refactored and checked that the refactorings were performance neutral.

Reply:
And it is faster than the final version?